parsers/c/cn_lexer.mll

```diff
@@ -0,0 +1,261 @@
+  ref None
+
+}
+
+(* ========================================================================== *)
+
+(* Integer constant - we follow C but omit suffixes (see ISO C11 §6.4.4.1#1) *)
+let digit = ['0'-'9']
+let hexadecimal_digit = ['0'-'9' 'A'-'F' 'a'-'f']
+let octal_digit = ['0'-'7']
+let nonzero_digit = ['1'-'9']
+
+let hexadecimal_prefix = "0x" | "0X"
+let hexadecimal_constant = hexadecimal_prefix hexadecimal_digit+
+
+(* C23 binary constant, but omitting ' separators for now *)
+(* TODO(Christopher/Dhruv): add support for ' separators to be in line with C23 ? *)
+let binary_prefix = "0b" | "0B"
+let binary_digit = "0" | "1"
+let binary_constant =
+  (* binary_prefix (binary_digit "'"?)+ *)
+  binary_prefix (binary_digit)+
+
+let octal_constant = '0' octal_digit*
+let decimal_constant = nonzero_digit digit*
+
+let integer_constant =
+    decimal_constant
+  | octal_constant
+  | hexadecimal_constant
+  | binary_constant
+
+(* Identifiers - we follow C11 *)
+(* TODO(Christopher/Dhruv): do you care about following C11 closely for this
+    (e.g. with respect to the universal character stuff), or should we simplify
+    like we already do for uppercase names. *)
+(* see ISO C11 §6.4.3#1 *)
+let hex_quad =
+  hexadecimal_digit hexadecimal_digit
+    hexadecimal_digit hexadecimal_digit
+
+let universal_character_name =
+    "\\u" hex_quad
+  | "\\U" hex_quad hex_quad
+
+(* see ISO C11 §6.4.4.1#1 *)
+let nondigit = ['_' 'a'-'z' 'A'-'Z']
+let identifier_nondigit =
+    nondigit
+  | universal_character_name
+let identifier = identifier_nondigit (identifier_nondigit | digit)*
+
+
+(* Whitespaces *)
+let whitespace_char = [' ' '\t' (*'\n'*) '\012' '\r']
+
+let integer_width = ("8" | "16" | "32" | "64" | "128")
+
+(* ========================================================================== *)
+
+rule onelinecomment = parse
+  | '\n' | eof { () }
+  | _          { lex_comment onelinecomment lexbuf }
+
+(* Entry point *)
+and initial = parse
+  (* Single-line comment *)
+  | "//" {let _ = onelinecomment lexbuf in Lexing.new_line lexbuf; initial lexbuf}
+
+  | '\n'             { Lexing.new_line lexbuf; initial lexbuf }
+  | whitespace_char+ { initial lexbuf }
+
+  | (integer_constant as str) 'u' (integer_width as n)
+      { BITS_CONSTANT (str, `U, int_of_string n) }
+  | (integer_constant as str) 'i' (integer_width as n)
+      { BITS_CONSTANT (str, `I, int_of_string n) }
+  | (integer_constant as str)
+      (* TODO(K): (!!!) when moving away froms Cabs.constant in this token, be careful about the
+          decoding of octal constants. Z.of_string won't do the right (C behaviour) thing. *)
+      { INTEGER_CONSTANT (Cabs.CabsInteger_const (str, None)) }
+
+  | '<' ([^'='][^'<' '>']+ as str) '>'
+      {
+        try
+          LT_CTYPE_GT (C_parser.type_name_eof C_lexer.lexer (Lexing.from_string str))
+        with
+          | _ ->
+              relexbuf_opt := Some (Lexing.from_string (str ^ ">"));
+              LT
+      }
+
```

Comment:
This is an amazing hack. But I do worry it might be a bit too hacky. Can it be more conservative? Specifically, I wanted to have `<-` available as a token. We do need to discuss this properly though @cp526

Reply:
The idea of `...<...>` was to have a C++-like syntax for type parameters, which would be good to keep if we can. What's the use of `<-`?

Reply:
I think types always start with a letter (or maybe `_`), so you could probably change it to only try for a type, if the `<` is followed by one of these. I think should allow you to have `<-` later if you wanted it.

Otoh, I don't think this would work if you ever wanted types that have instantiations in them, but we probably don't have those at the moment.